File: services/asset-core/internal/delivery/http/errorcodes.go
package http

import (
	"encoding/json"
	"net/http"

	"seta-im-intern/go-asset-core/internal/requestcontext"
)

type ErrorCode struct {
	Code    string
	Number  int
	Message string
}

var errorCodes = map[string]ErrorCode{
	"INTERNAL_ERROR":                {"INTERNAL_ERROR", 1000, "Internal server error, please try again"},
	"BAD_REQUEST":                   {"BAD_REQUEST", 1001, "Malformed request body or parameters"},
	"METHOD_NOT_ALLOWED":            {"METHOD_NOT_ALLOWED", 1002, "HTTP method not allowed on this endpoint"},
	"CURSOR_INVALID":                {"CURSOR_INVALID", 1003, "Pagination cursor is malformed or stale"},
	"UNAUTHENTICATED":               {"UNAUTHENTICATED", 2001, "Missing or invalid actor identity"},
	"NO_ORG_CONTEXT":                {"NO_ORG_CONTEXT", 2002, "Organization context is required"},
	"FORBIDDEN":                     {"FORBIDDEN", 2003, "The requested action is not permitted"},
	"USER_NOT_FOUND":                {"USER_NOT_FOUND", 2004, "User not found or inactive"},
	"UNKNOWN_ACTION":                {"UNKNOWN_ACTION", 2005, "Requested permission action is not recognized"},
	"TRAINER_BYPASS_DISABLED":       {"TRAINER_BYPASS_DISABLED", 2006, "Trainer bypass is disabled"},
	"TRAINER_BYPASS_EXPIRED":        {"TRAINER_BYPASS_EXPIRED", 2007, "Trainer bypass has expired"},
	"RESERVED_ROLE_CODE":            {"RESERVED_ROLE_CODE", 2008, "Role code is reserved and cannot be modified"},
	"FOLDER_NOT_FOUND":              {"FOLDER_NOT_FOUND", 3001, "Folder not found"},
	"FOLDER_ORG_MISMATCH":           {"FOLDER_ORG_MISMATCH", 3002, "Folder not found"},
	"FOLDER_NAME_CONFLICT":          {"FOLDER_NAME_CONFLICT", 3003, "A folder with this name already exists at this location"},
	"FOLDER_NOT_EMPTY":              {"FOLDER_NOT_EMPTY", 3004, "Folder contains active descendants or metadata"},
	"FOLDER_CYCLE_DETECTED":         {"FOLDER_CYCLE_DETECTED", 3005, "Move would create a cycle"},
	"FOLDER_PARENT_DELETED":         {"FOLDER_PARENT_DELETED", 3006, "Parent folder is deleted or missing"},
	"FOLDER_ROOT_PROTECTED":         {"FOLDER_ROOT_PROTECTED", 3007, "Cannot perform this action on the root folder"},
	"DELETION_PREVIEW_STALE":        {"DELETION_PREVIEW_STALE", 3008, "Folder deletion preview is stale; request a new preview"},
	"FOLDER_DELETION_IN_PROGRESS":   {"FOLDER_DELETION_IN_PROGRESS", 3009, "Folder deletion is already in progress"},
	"DELETION_JOB_NOT_FOUND":        {"DELETION_JOB_NOT_FOUND", 3010, "Folder deletion job not found"},
	"DELETION_JOB_NOT_CANCELLABLE":  {"DELETION_JOB_NOT_CANCELLABLE", 3011, "Folder deletion job cannot be cancelled or retried in its current state"},
	"FOLDER_NOT_DELETED":            {"FOLDER_NOT_DELETED", 3012, "Folder is already active"},
	"RESTORE_PARENT_DELETED":        {"RESTORE_PARENT_DELETED", 3013, "Restore the parent folder before restoring this item"},
	"LIFECYCLE_UNIT_NOT_FOUND":      {"LIFECYCLE_UNIT_NOT_FOUND", 3014, "Lifecycle unit not found"},
	"LIFECYCLE_UNIT_NOT_RESTORABLE": {"LIFECYCLE_UNIT_NOT_RESTORABLE", 3015, "Lifecycle unit cannot be restored in its current state"},
	"LIFECYCLE_JOB_NOT_FOUND":       {"LIFECYCLE_JOB_NOT_FOUND", 3016, "Lifecycle job not found"},
	"METADATA_NOT_FOUND":            {"METADATA_NOT_FOUND", 4001, "Metadata item not found"},
	"METADATA_IDENTITY_CONFLICT":    {"METADATA_IDENTITY_CONFLICT", 4002, "External identity already exists on an active item"},
	"METADATA_VALIDATION_ERROR":     {"METADATA_VALIDATION_ERROR", 4003, "Metadata field validation failed"},
	"METADATA_FOLDER_DELETED":       {"METADATA_FOLDER_DELETED", 4004, "Containing folder is deleted"},
	"METADATA_NOT_DELETED":          {"METADATA_NOT_DELETED", 4005, "Metadata item is already active"},
	"GRANT_NOT_FOUND":               {"GRANT_NOT_FOUND", 5001, "Object permission not found"},
	"GRANT_INVALID_TARGET":          {"GRANT_INVALID_TARGET", 5002, "Grant must target exactly one of user or role"},
	"MEDIA_UPLOAD_NOT_FOUND":        {"MEDIA_UPLOAD_NOT_FOUND", 6001, "Media upload not found"},
	"MEDIA_UPLOAD_IN_PROGRESS":      {"MEDIA_UPLOAD_IN_PROGRESS", 6002, "A media upload is already in progress for this asset"},
	"IDEMPOTENCY_KEY_REUSED":        {"IDEMPOTENCY_KEY_REUSED", 6003, "Idempotency key was already used with different request data"},
	"UPLOAD_SESSION_EXPIRED":        {"UPLOAD_SESSION_EXPIRED", 6004, "Upload session has expired"},
	"MEDIA_PAYLOAD_TOO_LARGE":       {"MEDIA_PAYLOAD_TOO_LARGE", 6005, "Media file exceeds the maximum allowed size"},
	"MEDIA_TYPE_UNSUPPORTED":        {"MEDIA_TYPE_UNSUPPORTED", 6006, "Media type is not supported"},
	"MEDIA_OBJECT_MISMATCH":         {"MEDIA_OBJECT_MISMATCH", 6007, "Stored object does not match the upload session"},
	"INVALID_IMAGE":                 {"INVALID_IMAGE", 6008, "File is not a valid image"},
	"IMAGE_DIMENSIONS_EXCEEDED":     {"IMAGE_DIMENSIONS_EXCEEDED", 6009, "Image dimensions exceed the maximum allowed"},
	"PROCESSING_TIMEOUT":            {"PROCESSING_TIMEOUT", 6010, "Media processing exceeded its time limit"},
	"MEDIA_STORAGE_UNAVAILABLE":     {"MEDIA_STORAGE_UNAVAILABLE", 6011, "Media storage is temporarily unavailable"},
	"MEDIA_PROCESSING_FAILED":       {"MEDIA_PROCESSING_FAILED", 6012, "Media processing failed"},
	"MEDIA_QUOTA_EXCEEDED":          {"MEDIA_QUOTA_EXCEEDED", 6013, "Organization media storage quota exceeded"},
	"MEDIA_RATE_LIMITED":            {"MEDIA_RATE_LIMITED", 6014, "Too many upload requests, please retry later"},
	"MEDIA_UPLOAD_STATE_CONFLICT":   {"MEDIA_UPLOAD_STATE_CONFLICT", 6015, "Operation is not valid for the current upload state"},
}

// internalMediaErrorCodes never reach a client. Notification isolation is an
// operator-facing state, so it is recorded and mapped to a safe public code
// rather than being added to the shared table.
var internalMediaErrorCodes = map[string]string{
	"MEDIA_NOTIFICATION_ISOLATED": "MEDIA_PROCESSING_FAILED",
}

// publicMediaErrorCode maps an internal media failure category to the safe code
// a client may see. An unrecognised category degrades to the generic media
// processing failure rather than leaking its own name.
func publicMediaErrorCode(internalCode string) string {
	if publicCode, ok := internalMediaErrorCodes[internalCode]; ok {
		return publicCode
	}
	if _, ok := errorCodes[internalCode]; ok {
		return internalCode
	}
	return "MEDIA_PROCESSING_FAILED"
}

func lookupErrorCode(code string) ErrorCode {
	if definition, ok := errorCodes[code]; ok {
		return definition
	}
	return errorCodes["INTERNAL_ERROR"]
}

func writeError(w http.ResponseWriter, r *http.Request, status int, code string) {
	definition := lookupErrorCode(code)
	requestcontext.RecordError(r.Context(), definition.Code, definition.Number)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(map[string]any{
		"error": map[string]any{
			"code":    definition.Code,
			"number":  definition.Number,
			"message": definition.Message,
			"traceId": requestcontext.TraceID(r.Context()),
			"service": ServiceNameAssetCore,
		},
	})
}

// writeLegacyError is a temporary classifier for handlers that still return
// legacy literals. It always delegates to writeError, the single safe-envelope
// emitter, until those handlers are migrated to explicit error codes.
func writeLegacyError(w http.ResponseWriter, r *http.Request, message string, status int) {
	code := "INTERNAL_ERROR"
	switch message {
	case "Method not allowed":
		code = "METHOD_NOT_ALLOWED"
	case "invalid internal service credential", "missing X-User-Id or X-Org-Id header":
		code = "UNAUTHENTICATED"
	case "Organization context mismatch":
		code = "FORBIDDEN"
	case "Folder not found":
		code = "FOLDER_NOT_FOUND"
	case "Invalid input":
		code = "BAD_REQUEST"
	default:
		if status == http.StatusBadRequest {
			code = "BAD_REQUEST"
		}
	}
	writeError(w, r, status, code)
}

File: services/access-core/src/errors/errorCodes.ts
export interface ErrorDefinition {
  code: string;
  number: number;
  message: string;
}

export const errorDefinitions: ErrorDefinition[] = [
  { code: "INTERNAL_ERROR", number: 1000, message: "Internal server error, please try again" },
  { code: "BAD_REQUEST", number: 1001, message: "Malformed request body or parameters" },
  { code: "METHOD_NOT_ALLOWED", number: 1002, message: "HTTP method not allowed on this endpoint" },
  { code: "CURSOR_INVALID", number: 1003, message: "Pagination cursor is malformed or stale" },
  { code: "UNAUTHENTICATED", number: 2001, message: "Missing or invalid actor identity" },
  { code: "NO_ORG_CONTEXT", number: 2002, message: "Organization context is required" },
  { code: "FORBIDDEN", number: 2003, message: "The requested action is not permitted" },
  { code: "USER_NOT_FOUND", number: 2004, message: "User not found or inactive" },
  { code: "UNKNOWN_ACTION", number: 2005, message: "Requested permission action is not recognized" },
  { code: "TRAINER_BYPASS_DISABLED", number: 2006, message: "Trainer bypass is disabled" },
  { code: "TRAINER_BYPASS_EXPIRED", number: 2007, message: "Trainer bypass has expired" },
  { code: "RESERVED_ROLE_CODE", number: 2008, message: "Role code is reserved and cannot be modified" },
  { code: "FOLDER_NOT_FOUND", number: 3001, message: "Folder not found" },
  { code: "FOLDER_ORG_MISMATCH", number: 3002, message: "Folder not found" },
  { code: "FOLDER_NAME_CONFLICT", number: 3003, message: "A folder with this name already exists at this location" },
  { code: "FOLDER_NOT_EMPTY", number: 3004, message: "Folder contains active descendants or metadata" },
  { code: "FOLDER_CYCLE_DETECTED", number: 3005, message: "Move would create a cycle" },
  { code: "FOLDER_PARENT_DELETED", number: 3006, message: "Parent folder is deleted or missing" },
  { code: "FOLDER_ROOT_PROTECTED", number: 3007, message: "Cannot perform this action on the root folder" },
  { code: "DELETION_PREVIEW_STALE", number: 3008, message: "Folder deletion preview is stale; request a new preview" },
  { code: "FOLDER_DELETION_IN_PROGRESS", number: 3009, message: "Folder deletion is already in progress" },
  { code: "DELETION_JOB_NOT_FOUND", number: 3010, message: "Folder deletion job not found" },
  { code: "DELETION_JOB_NOT_CANCELLABLE", number: 3011, message: "Folder deletion job cannot be cancelled or retried in its current state" },
  { code: "METADATA_NOT_FOUND", number: 4001, message: "Metadata item not found" },
  { code: "METADATA_IDENTITY_CONFLICT", number: 4002, message: "External identity already exists on an active item" },
  { code: "METADATA_VALIDATION_ERROR", number: 4003, message: "Metadata field validation failed" },
  { code: "METADATA_FOLDER_DELETED", number: 4004, message: "Containing folder is deleted" },
  { code: "GRANT_NOT_FOUND", number: 5001, message: "Object permission not found" },
  { code: "GRANT_INVALID_TARGET", number: 5002, message: "Grant must target exactly one of user or role" },
];

const byCode = new Map(errorDefinitions.map((definition) => [definition.code, definition]));

const legacyAliases: Record<string, string> = {
  BAD_USER_INPUT: "BAD_REQUEST",
  INTERNAL_SERVER_ERROR: "INTERNAL_ERROR",
};

export function getErrorDefinition(code: unknown): ErrorDefinition {
  const normalized = typeof code === "string" ? (legacyAliases[code] ?? code) : "";
  return byCode.get(normalized) ?? byCode.get("INTERNAL_ERROR")!;
}

export function isKnownErrorCode(code: unknown): code is string {
  return typeof code === "string" && (byCode.has(code) || code in legacyAliases);
}
